File: Monthly Challenges/September 2020 Monthly Challenge/Day18.java
public class Day18 {
    public int maxProfit(int[] prices) {
        if (prices.length < 2) {
            return 0;
        }
        
        int profit = 0;
        int curr_min = prices[0];
        
        for (int i = 1; i < prices.length; ++i) {
            profit = Math.max(profit, prices[i] - curr_min);
            curr_min = Math.min(curr_min, prices[i]);
        }
        
        return profit;
    }
}

File: Monthly Challenges/October 2020 Monthly Challenge/Day18.scala
object Day18 {
    def maxProfit(k: Int, prices: Array[Int]): Int = {
        if (k == 0 || prices.length < 2) return 0
        val max_transactions = Math.min(k, prices.length / 2);

        val purchases : Array[Int] = Array.fill(max_transactions)(-prices(0))
        val sales : Array[Int] = new Array[Int](max_transactions)
        
        for (i <- 1 until prices.length) {
            purchases(0) = math.max(purchases(0), -prices(i))
            sales(0) = math.max(sales(0), purchases(0) + prices(i))
            
            for (j <- 1 until max_transactions) {
                purchases(j) = math.max(purchases(j), sales(j - 1) - prices(i))
                sales(j) = math.max(sales(j), purchases(j) + prices(i))
            }
        }
        
        sales(max_transactions - 1)
    }
}
